internal/services/serviceprincipals/synchronization_secret_resource.go

```diff
@@ -0,0 +1,217 @@
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"key": {
+							Description: "Name for this key-value pair.",
+							Type:        schema.TypeString,
+							Required:    true,
+						},
+						"value": {
+							Description: "Value for this key-value pair.",
+							Type:        schema.TypeString,
+							Required:    true,
+							Sensitive:   true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func synchronizationSecretResourceCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*clients.Client).ServicePrincipals.SynchronizationJobClient
+	spClient := meta.(*clients.Client).ServicePrincipals.ServicePrincipalsClient
+	objectId := d.Get("service_principal_id").(string)
+
+	tf.LockByName(servicePrincipalResourceName, objectId)
+	defer tf.UnlockByName(servicePrincipalResourceName, objectId)
+
+	servicePrincipal, status, err := spClient.Get(ctx, objectId, odata.Query{})
+	if err != nil {
+		if status == http.StatusNotFound {
+			return tf.ErrorDiagPathF(nil, "service_principal_id", "Service principal with object ID %q was not found", objectId)
+		}
+		return tf.ErrorDiagPathF(err, "service_principal_id", "Retrieving service principal with object ID %q", objectId)
+	}
+	if servicePrincipal == nil || servicePrincipal.ID == nil {
+		return tf.ErrorDiagF(errors.New("nil service principal or service principal with nil ID was returned"), "API error retrieving service principal with object ID %q", objectId)
+	}
+
+	synchronizationSecret := msgraph.SynchronizationSecret{
+		Credentials: expandSynchronizationSecretKeyStringValuePair(d.Get("credentials").([]interface{})),
+	}
+
+	_, err = client.SetSecrets(ctx, synchronizationSecret, *servicePrincipal.ID)
+	if err != nil {
+		return tf.ErrorDiagF(err, "Creating synchronization secret for service principal ID %q", *servicePrincipal.ID)
+	}
+	id := parse.NewSynchronizationSecretID(*servicePrincipal.ID)
+
+	// Wait for the secret to appear
+	timeout, _ := ctx.Deadline()
+	_, err = (&resource.StateChangeConf{
+		Pending:                   []string{"Waiting"},
+		Target:                    []string{"Done"},
+		Timeout:                   time.Until(timeout),
+		MinTimeout:                1 * time.Second,
+		ContinuousTargetOccurence: 5,
+		Refresh: func() (interface{}, string, error) {
+			newSynchronizationSecret, _, err := client.GetSecrets(ctx, id.ServicePrincipalId)
+			if err != nil {
+				return nil, "Error", fmt.Errorf("retrieving member")
+			}
+			if newSynchronizationSecret != nil {
+				if len(*synchronizationSecret.Credentials) == len(*newSynchronizationSecret.Credentials) {
+					return "stub", "Done", nil
+				}
+				return "stub", "Waiting", nil
+			} else {
+				return "stub", "Waiting", nil
+			}
+		},
+	}).WaitForStateContext(ctx)
+
+	if err != nil {
+		return tf.ErrorDiagF(err, "Waiting for synchronization secret %q", id.ServicePrincipalId)
+	}
+
+	d.SetId(id.String())
```

Comment:
Since this is used for updating too, this should only be set on creation
```suggestion
	if d.IsNewResource() {
		d.SetId(id.String())
	}
```